app/src/main/java/com/asfoundation/wallet/ui/iab/BillingWebViewFragment.kt

```diff
@@ -0,0 +1,179 @@
+      override fun shouldOverrideUrlLoading(view: WebView,
+                                            clickUrl: String): Boolean {
+        if (clickUrl.contains(
+                LOCAL_PAYMENTS_SCHEMA) || clickUrl.contains(
+                ADYEN_PAYMENT_SCHEMA)) {
+          currentUrl = clickUrl
+          finishWithSuccess(clickUrl)
+        } else if (clickUrl.contains(
+                GO_PAY_APP_PAYMENTS_SCHEMA) || clickUrl.contains(
+                LINE_APP_PAYMENTS_SCHEMA)) {
+          launchActivity(Intent(Intent.ACTION_VIEW, Uri.parse(clickUrl)))
+        } else if (clickUrl.contains(
+                CODAPAY_FINAL_REDIRECT_SCHEMA) && clickUrl.contains(
+                ORDER_ID_PARAMETER)) {
+          val orderId = Uri.parse(clickUrl)
+              .getQueryParameter(ORDER_ID_PARAMETER)
+          finishWithSuccess(LOCAL_PAYMENTS_URL + orderId)
+        } else {
+          currentUrl = clickUrl
+          return false
+        }
+        return true
+      }
+
+      override fun shouldOverrideUrlLoading(view: WebView,
+                                            request: WebResourceRequest): Boolean {
+        return super.shouldOverrideUrlLoading(view, request)
+      }
+
+      override fun onPageFinished(view: WebView, url: String) {
+        super.onPageFinished(view, url)
+        if (!url.contains("/redirect")) {
+          val timeout = timeoutReference.getAndSet(null)
+          timeout?.cancel(false)
+          webview_progress_bar?.visibility = View.GONE
+          webview.visibility = View.VISIBLE
+        }
+        if (url.contains(ASYNC_PAYMENT_FORM_SHOWN_SCHEMA)) {
+          asyncDetailsShown = true
+        }
+      }
+    }
+    view.webview.settings.javaScriptEnabled = true
+    view.webview.loadUrl(currentUrl)
+    return view
+  }
+
+  fun handleBackPressed(): Boolean {
+    return if (asyncDetailsShown) {
+      webview.loadUrl("https://pay.dlocal.com/payment_method_connectors/global_pm//back")
+      asyncDetailsShown = false
+      true
+    } else {
+      false
+    }
+  }
+
+  override fun onSaveInstanceState(outState: Bundle) {
+    super.onSaveInstanceState(outState)
+    outState.putString(CURRENT_URL, currentUrl)
+  }
+
+  override fun onDestroy() {
+    executorService!!.shutdown()
+    super.onDestroy()
+  }
+
+  override fun onDetach() {
+    webViewActivity = null
+    webview?.webViewClient = null
+    super.onDetach()
+  }
+
+  private fun launchActivity(intent: Intent) {
+    try {
+      startActivity(intent)
+    } catch (exception: ActivityNotFoundException) {
+      exception.printStackTrace()
+      if (view != null) {
+        Snackbar.make(view!!, R.string.unknown_error,
+            Snackbar.LENGTH_SHORT)
+            .show()
+      }
+    }
+  }
+
+  private fun finishWithSuccess(url: String) {
+    val intent = Intent()
+    intent.data = Uri.parse(url)
+    webViewActivity!!.setResult(WebViewActivity.SUCCESS, intent)
+    webViewActivity!!.finish()
+  }
+
+  companion object {
```

Comment:
I think we had agreed that companion object should be placed on top of class files right?

Reply:
Correct